### PR TITLE
[WIP] feat(docs):Addition of JS/TS Toggle button in the docs to switch to appropriate code.

### PR DESCRIPTION
**Pull request by noopurphalak**
### Before submitting the PR, please make sure you do the following

- [x] Read the [Contributing Guidelines](https://github.com/vueuse/vueuse/blob/main/CONTRIBUTING.md).
- [x] Read the [Pull Request Guidelines](https://github.com/vueuse/vueuse/blob/main/packages/guidelines.md).
- [x] Check that there isn't already a PR that solves the problem the same way to avoid creating a duplicate.
- [x] Provide a description in this PR that addresses **what** the PR is solving, or reference the issue that it solves (e.g. `fixes #123`).
- [x] Ideally, include relevant tests that fail without this PR but pass with it.

<details>
<summary><strong>⚠️ Slowing down new functions</strong></summary>
<br>

> **Warning**: **Slowing down new functions**
>
> As the VueUse audience continues to grow, we have been inundated with an overwhelming number of feature requests and pull requests. As a result, maintaining the project has become increasingly challenging and has stretched our capacity to its limits. As such, in the near future, we may need to slow down our acceptance of new features and prioritize the stability and quality of existing functions. **Please note that new features for VueUse may not be accepted at this time.** If you have any new ideas, we suggest that you first incorporate them into your own codebase, iterate on them to suit your needs, and assess their generalizability. If you strongly believe that your ideas are beneficial to the community, you may submit a pull request along with your use cases, and we would be happy to review and discuss them. Thank you for your understanding.

</details>

---

### Description

Resolves https://github.com/vueuse/vueuse/issues/2475

### Additional context

This PR introduces a toggle for JS/TS switch on every function page in the vueuse documentation.

This PR is still work in progress, I will mark it for review as soon as it is ready for review. 

---

### <samp>🤖 Generated by Copilot at 425a218</samp>

This pull request adds a feature to the documentation site that allows users to switch between TypeScript and JavaScript code examples. It introduces a new dependency `@vue/theme` and a new component `TsJsToggleSwitch.vue` for the sidebar toggle. It also updates the documentation page for `createGlobalState` and the `pnpm-lock.yaml` file.

### <samp>🤖 Generated by Copilot at 425a218</samp>

*  Add the `@vue/theme` dependency to enable a custom theme for the Vitepress documentation site ([link](https://github.com/vueuse/vueuse/pull/3082/files?diff=unified&w=0#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519R55), [link](https://github.com/vueuse/vueuse/pull/3082/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbR14-R17), [link](https://github.com/vueuse/vueuse/pull/3082/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbR4019-R4038))
*  Import and use the `TsJsToggleSwitch` component to render a language preference switch in the sidebar ([link](https://github.com/vueuse/vueuse/pull/3082/files?diff=unified&w=0#diff-8ac7c8298153b8b006a414a6ff0abde634bd9a6483eec29d42ad4387981aa170R1-R198), [link](https://github.com/vueuse/vueuse/pull/3082/files?diff=unified&w=0#diff-7419976a7494d31468f916e0471f372f1dbcc8c98250256beae2e20a34f3cefaL1-R3), [link](https://github.com/vueuse/vueuse/pull/3082/files?diff=unified&w=0#diff-7419976a7494d31468f916e0471f372f1dbcc8c98250256beae2e20a34f3cefaR15-R19), [link](https://github.com/vueuse/vueuse/pull/3082/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL4046-R4068), [link](https://github.com/vueuse/vueuse/pull/3082/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL4059-R4085))
*  Add `<div class="js-api">` and `<div class="ts-api">` tags to the documentation pages to toggle the visibility of JavaScript and TypeScript code examples ([link](https://github.com/vueuse/vueuse/pull/3082/files?diff=unified&w=0#diff-7f67cdcb8dcdeface98a1bc3d75c2b1863a7172c62636356913991117e473cb5R14-R15), [link](https://github.com/vueuse/vueuse/pull/3082/files?diff=unified&w=0#diff-7f67cdcb8dcdeface98a1bc3d75c2b1863a7172c62636356913991117e473cb5R29-R48))
*  Remove the `dev: true` property from several dependencies in the lock file that are also used for production ([link](https://github.com/vueuse/vueuse/pull/3082/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL413), [link](https://github.com/vueuse/vueuse/pull/3082/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL424-R429), [link](https://github.com/vueuse/vueuse/pull/3082/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL434-R437), [link](https://github.com/vueuse/vueuse/pull/3082/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL444), [link](https://github.com/vueuse/vueuse/pull/3082/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL452), [link](https://github.com/vueuse/vueuse/pull/3082/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL461), [link](https://github.com/vueuse/vueuse/pull/3082/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL468), [link](https://github.com/vueuse/vueuse/pull/3082/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL476), [link](https://github.com/vueuse/vueuse/pull/3082/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL484-R480), [link](https://github.com/vueuse/vueuse/pull/3082/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL494), [link](https://github.com/vueuse/vueuse/pull/3082/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL500-R493), [link](https://github.com/vueuse/vueuse/pull/3082/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL510), [link](https://github.com/vueuse/vueuse/pull/3082/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL518), [link](https://github.com/vueuse/vueuse/pull/3082/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL1835), [link](https://github.com/vueuse/vueuse/pull/3082/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL1847), [link](https://github.com/vueuse/vueuse/pull/3082/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL1869), [link](https://github.com/vueuse/vueuse/pull/3082/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL3446), [link](https://github.com/vueuse/vueuse/pull/3082/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL4169), [link](https://github.com/vueuse/vueuse/pull/3082/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL4194), [link](https://github.com/vueuse/vueuse/pull/3082/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL8315), [link](https://github.com/vueuse/vueuse/pull/3082/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL8559), [link](https://github.com/vueuse/vueuse/pull/3082/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL9902), [link](https://github.com/vueuse/vueuse/pull/3082/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL10680), [link](https://github.com/vueuse/vueuse/pull/3082/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL12059), [link](https://github.com/vueuse/vueuse/pull/3082/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL12167-R12194))
*  Update the `body-scroll-lock` dependency to version `4.0.0-beta.0` and add version `3.1.5` as a dependency of the `@vue/theme` package ([link](https://github.com/vueuse/vueuse/pull/3082/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL4468-R4496))
*  Add the `normalize.css` dependency to the `@vue/theme` package and remove it from the `dev: true` dependencies ([link](https://github.com/vueuse/vueuse/pull/3082/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbR8959-R8962), [link](https://github.com/vueuse/vueuse/pull/3082/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL8315))
*  Remove an empty line from the `createGlobalState` documentation page for readability ([link](https://github.com/vueuse/vueuse/pull/3082/files?diff=unified&w=0#diff-7f67cdcb8dcdeface98a1bc3d75c2b1863a7172c62636356913991117e473cb5L52))

### CODE REVIEW

**Comment by noopurphalak** [comment]
@antfu I have create a toggle button on the top of the sidebar in this PR. The PR is still in progress as now I have to write all the switchable content for all the functions. I wanted some clarification from you on some thing.

In the PR, I have written the typescript code snippet for `createGlobalState`  as follows:

```typescript
// store.ts
import { ref } from 'vue'
import { createGlobalState } from '@vueuse/core'
export const useGlobalState = createGlobalState(
  () => {
    const count = ref(0)
    return { count }
  }
)
```

This is same as the javascript code. This is because all the types are already inferred from the `createGlobalState` function in the output. Is this okay or should I write the function as below:

```typescript
// store.ts
import { ref } from 'vue'
import { createGlobalState } from '@vueuse/core'

const storageFn =  () => {
    const count = ref(0)
    return { count }
}
export const useGlobalState = createGlobalState<typeof storageFn>(storageFn)
```

**Comment by antfu** [comment]
Thanks!

I think we should not manually write js/ts docs version **at all**. Instead, as I suggested (https://github.com/vueuse/vueuse/issues/2475#issuecomment-1326683188), we should always write in TS, and use a custom plugin to generate the js version from the ts blocks using esbuild at compile time

**Comment by noopurphalak** [comment]
Thanks for your reply @antfu  .  I am actually not aware as to how to do that. Can you may be give ma an example for one of the functions so that I have an idea of how to do it?

**Comment by noopurphalak** [comment]
@antfu Can you let me know how we can debug the vite plugin in this project? I need to see if the code is getting filtered correctly for transformation...

**Comment by noopurphalak** [comment]
Hey @antfu , As suggested by you, I have created a plugin that will create a JS frame for the typescript snippet it finds in all the `index.md` files. Can you review it once? I can then change the JS snippets in every function to show typescript by default and they can be transpiled to JS frames on build time?

**Comment by noopurphalak** [comment]
@antfu Request you to please check if the plugin I created for typescript transpilation to javascript is what we require? I can then move ahead with converting code snippets to typescript for other functions...

File: packages/shared/createGlobalState/index.md [hunk 1]
```diff
@@ -11,19 +11,24 @@ Keep states in the global scope to be reusable across Vue instances.
 
 ### Without Persistence (Store in Memory)
 
-```js
-// store.js
+<div class="ts-api">
```
Comment (antfu) [inline packages/shared/createGlobalState/index.md:14]: Can we just automatically apply this to any ts block, without the need of this marker? (so the documents keep untouched)

Comment (noopurphalak) [reply packages/shared/createGlobalState/index.md:14]: the `<div class="ts-api">...</div>` and `<div class="js-api">...</div>` are identifier classes for when to hide which code block when the toggle is used.

Please refer the below code in `TsJsToggleSwitch.vue`:

https://github.com/vueuse/vueuse/pull/3082/files#diff-8ac7c8298153b8b006a414a6ff0abde634bd9a6483eec29d42ad4387981aa170R174-R184

What `TsJsToggleSwitch.vue` does is, the default code enabled is the Typescript code, when we toggle to Javascript, the component adds a `prefer-js` class to the body. Once this class is attached to the body, the typescript code (one with class `ts-api`) gets a `display: none;` css property and the Javascript code (one with class `js-api`) gets a `display: initial;` css property. Similarly, when you toggle it back to typescript, the Javascript code gets removed from the DOM and the Typescript code gets added back to the DOM. Hence the `div` elements in the documents.

I have also written a vite plugin `tsToJsTransform.ts`, which recognizes the typescript code using the divs with the `ts-api` and transpiles the code inside it to Javascript code and adds both the snippets in the file code at build time. Correct me if I am wrong, but we are supposed to rewrite all the JS snippets as TS snippets and transpile those snippets on build time to generate an additional JS code block.

If you have a better approach, please let me know, and I shall try to go that way.

Comment (noopurphalak) [reply packages/shared/createGlobalState/index.md:14]: @antfu Please let me know if you have a better approach...

Comment (noopurphalak) [reply packages/shared/createGlobalState/index.md:14]: @antfu Can you please update on my comment?

**Pull request closed without merging**